neo/Consensus/ConsensusContext.cs

```diff
@@ -106,25 +119,41 @@ public void Deserialize(BinaryReader reader)
                     Commits[i] = null;
             }
             ExpectedView = reader.ReadVarBytes();
+            ChangeViewWitnessInvocationScripts = new byte[reader.ReadVarInt()][];
+            for (int i = 0; i < ChangeViewWitnessInvocationScripts.Length; i++)
+            {
+                ChangeViewWitnessInvocationScripts[i] = reader.ReadVarBytes();
+                if (ChangeViewWitnessInvocationScripts[i].Length == 0)
+                    ChangeViewWitnessInvocationScripts[i] = null;
+            }
+            ChangeViewTimestamps = reader.ReadSerializableArray(Validators.Length);
         }
 
         public void Dispose()
         {
             snapshot?.Dispose();
         }
 
-        public uint GetPrimaryIndex(byte view_number)
+        public uint GetPrimaryIndex(byte viewNumber)
         {
-            int p = ((int)BlockIndex - view_number) % Validators.Length;
+            int p = ((int)BlockIndex - viewNumber) % Validators.Length;
             return p >= 0 ? (uint)p : (uint)(p + Validators.Length);
         }
 
         public ConsensusPayload MakeChangeView()
         {
-            return MakeSignedPayload(new ChangeView
+            var payload = MakeSignedPayload(new ChangeView
             {
+                // View number is ignored for ChangeView, set it to 0, so it doesn't have to be saved for recovery.
+                ViewNumber = 0,
```

Comment:
In the original OnChangeViewReceived method:

https://github.com/neo-project/neo/blob/7cf286941f7f5541e2ab4bf182583452eec8fc5a/neo/Consensus/ConsensusService.cs#L137-L144

We can easily see through the log what the original view number of the node requesting the changing view is.

But if you set the `ViewNumber` to `0` here, there is no way to see the original view number.

Reply:
Original view number for change view always counts up by `1`. If `NewViewNumber` is `2` the original view number had to be `1`, right?

Reply:
Not really. If the view number is `0`, when timeout, usually you will request to change view number to `1`. But if there is no successful change, and the new timeout expires, then it will continue to request to change the view number to `2`, the original view number is still `0`, will not change.

Reply:
Of course you are right. What use is the original view number though; none of the code uses it at all currently? I could save it and restore in `RecoveryMessage`, but is it really very useful?

Reply:
It is only 1 byte per validator, so I don't mind adding something to save it also. Would you like me to do that, even though we aren't using it currently?

Reply:
Currently, the only place where the original view number is used is the logging in `OnChangeViewReceived`. This log is useful for finding consensus faults.

Reply:
Ok, this is reasonable. I'll save and restore it from the recovery message.